core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala

```diff
@@ -0,0 +1,259 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util
+
+import java.io.FileNotFoundException
+
+import scala.collection.mutable
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+import org.apache.hadoop.fs.viewfs.ViewFileSystem
+import org.apache.hadoop.hdfs.DistributedFileSystem
+
+import org.apache.spark._
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.internal.Logging
+import org.apache.spark.metrics.source.HiveCatalogMetrics
+
+/**
+ * :: DeveloperApi ::
+ * Utility functions to simplify and speed-up file listing.
+ */
+@DeveloperApi
+object HadoopFSUtils extends Logging {
+  /**
+   * :: DeveloperApi ::
+   * Lists a collection of paths recursively. Picks the listing strategy adaptively depending
+   * on the number of paths to list.
+   *
+   * This may only be called on the driver.
+   *
+   * @return for each input path, the set of discovered files for the path
+   */
+  @DeveloperApi
+  def parallelListLeafFiles(sc: SparkContext, paths: Seq[Path], hadoopConf: Configuration,
```

Comment:
If you target to expose this to Java users as well, we shouldn't use Scala friendly collections like `Seq` or `Option`.

Reply:
Since we need to pass `SparkContext` into this method, why not add it to SparkContext and return an RDD instead? Then the user can use RDD APIs to do other work in parallel, for example, add a filter to avoid pulling all paths to the driver.

Reply:
Thanks. I'll address this in a separate PR. 

I'm not sure about adding this to RDD API though: the things done are very specific, such as handling `FileNotFoundException`, dealing with locality, threshold on whether to enable/disable parallelism etc. It is composed of basic API calls like `parallelize` and `mapPartitions`, and I'm not sure which part of this is general enough to be extracted as a separate RDD API.